### ISSUE TITLE
Implement an asynchronous BFT algorithm

### ISSUE THREAD

**Issue by kayabaNerve** [serai-dex/serai#333]
HoneyBadgerBFT (https://eprint.iacr.org/2016/199.pdf) is a BFT protocol which notably *can be run over Tor*. If we ever want to hide the IP addresses of multisig participants, we would need to move Tributaries over Tor which Tendermint would not be viable for. An asynchronous BFT algorithm, such as HoneyBadgerBFT, would be viable.

Since HoneyBadgerBFT, the field has been improved with:
- Aleph https://arxiv.org/pdf/1908.05156.pdf
- Dumbo https://eprint.iacr.org/2020/841
- Speeding Dumbo https://eprint.iacr.org/2022/027
- Dory https://eprint.iacr.org/2022/1709.pdf

I cannot comment on which is preferable at this time.

HoneyBadgerBFT specifically requires a trusted setup or synchronously performed DKG. Accordingly, we'd have to perform a DKG over Serai itself (preferably with off-chain blobs) to set up each Tributary.

Tributary itself would also have to be rewritten over a TX stream.

Supersedes #276, complicates #286, see #271 for a discussion of some relevance.

**Comment by kayabaNerve** [comment]
Narwhal/Tusk would also be a candidate: https://arxiv.org/abs/2105.11827

**Comment by kayabaNerve** [comment]
We technically don't need an asynchronous BFT algorithm, solely a partially synchronous BFT algorithm.

A recent such creation is [Motorway](https://arxiv.org/abs/2401.10369).

**Comment by kayabaNerve** [comment]
https://eprint.iacr.org/2024/142 has the lowest round complexity thus-far.

**Comment by kayabaNerve** [comment]
https://eprint.iacr.org/2024/653 is a new work on the matter.

**Comment by kayabaNerve** [comment]
https://eprint.iacr.org/2024/1235 on the liveness of Tusk and a necessary modification to prevent their posited attack.

**Comment by kayabaNerve** [comment]
My criteria for which we move forward with will probably whichever is simplest to understand and audit so long as it doesn't have unbearable performance. Tributaries are fine with latency on the scale of a few minutes so long as we have sufficient throughput.

**Comment by kayabaNerve** [comment]
AleoBFT, derived from Narwhal and Tusk, has work being done on formal verification. https://provable.com/blog/creating-aleobft-formal-verification-milestone

Due to how many of these protocols have subtle flaws, this is greatly respected and appreciated.

**Comment by kayabaNerve** [comment]
If we adopt any scheme requiring cryptography past a signature, we should consider how well it will hold up into a future with quantum computers. The current Tendermint implementation (and Serai itself) can quite quickly be modified to use Dilithium/NTRU-Prime.

**Comment by kayabaNerve** [comment]
https://github.com/serai-dex/serai/issues/395#issuecomment-2692105012

**Comment by binarybaron** [comment]
Is this something that is planned for the mainnet release? What are the assumptions regarding latency with the current approach?

We have written a LibP2P Tor transport here in case this is interesting to you: https://github.com/eigenwallet/core/tree/master/libp2p-tor

**Comment by kayabaNerve** [comment]
Definitely not. It'd take months and is unnecessary for a functional Serai. It's primarily necessary for enabling validators to run over anonymity networks. It's secondarily _potentially_ a latency improvement as the Serai coordinator currently only achieves agreement every six seconds (upon success), but this would be as progress occurs incurs (which may be less than six seconds or more). It's tertiarily a way to ensure a cohesive solution across Serai, potentially a post-quantum solution, whereas right now we're fragmented across our Tendermint implementation along with Parity's BABE/GRANDPA implementation.

**Comment by spirobel** [comment]
https://drive.google.com/file/d/1y_7ddr8oNOknTQYHzXeeMD2ProQ0WjMs/view alpenglow paper claims to achieve fastest possible consensus. will be the future consensus algorithm of solana

**Comment by kayabaNerve** [comment]
Alpenglow is partially-synchronous and only supports <20% being malicious (not <33%).

**Comment by spirobel** [comment]
20% + 20% offline
can you expand a bit on partial synchronicity? 
you want to avoid synchronicity to avoid having to wait when there is high latency on a node  / the connection is interrupted because of tor?

**Comment by kayabaNerve** [comment]
That's the limit on what they support. An adversary with 21% malicious can still violate safety.

Yes, the desire for asynchronous network is so consensus always progresses regardless of networking delays, assuming nodes can talk to each other in any way.

**Comment by spirobel** [comment]
their reasoning for picking this model:

> in a real large scale distributed blockchain system, we will probably
see significantly less than 33% byzantines. Instead, realistic bad behavior
often comes from machine misconfigurations, software bugs, and network or
power outages. In other words, large scale faults are likely accidents rather
than coordinated attacks.
This attack model paradigm shift opens an opportunity to reconsider the
classic 3f + 1 bound. Alpenglow is based on the 5f + 1 bound that has been

>introduced by Martin and Alvisi [MA06]. While being less tolerant to ortho-
dox byzantine attacks, the 5f + 1 bound offers other advantages [Von+24].

>Two rounds of voting are required for finalization if the adversary is strong.
However, if the adversary possesses less stake, or does not misbehave all the
time, it is possible for a correct 5f + 1 protocol to finalize a block in just a
single round of voting. [...]

offers the advantage to tolerate higher amount of node outages:
>However, if the network is not being attacked, or the adversary does not
leverage some network advantage, **Alpenglow can tolerate an even higher share
of nodes that simply crash.**

if the network gets disrupted (or becomes asynchronous if you will) it will halt and recover:
>As we explain later, Alpenglow is partially-synchronous, and Assumption 1
is enough to ensure that even an adversary completely controlling the network
(inspecting, delaying, and scheduling communication between correct nodes
at will) cannot violate safety. A network outage or partition would simply
cause the protocol to pause and continue as soon as communication is restored,
without any incorrect outcome.

in summary you get faster finality in the default case of a working network and there is a path to recovery in case there is an interruption.

In general I would focus on the work that comes out of the eth, sol  (and in theory sui) ecosystems. Everything else seems fringe, theoretical / behind the curve.

**Comment by kayabaNerve** [comment]
I read the reasoning. IMO, sub-optimal safety thresholds aren't worth the performance benefit. Additionally, making progress under synchrony isn't worth it when the network frequently isn't synchronous, hence the goal of an algorithm achieving asynchronous liveness.

There are algorithms which will attempt one round of a synchronously live protocol, falling back to an asynchronous protocol on failure, optimistically using a partially-synchronous algorithm before pessimistically making progress anyways (not just when sufficient synchrony is restored). It's not that this research could never be applicable here. It's that the priority is a safe algorithm comprehensive to asynchronicity, as usable to justify running validators behind networks with extremely-variable latency. Only after that core should we discuss increasing complexity to increase performance.

I respect Mysticeti. That doesn't change the algorithm I've cited as my preference is straightforward (simplicity is valuable to initially achieve goals, optimizations coming after) _and post-quantum_, meaning it'll stand arbitrarily long: not just the next three-to-five years. That isn't true for most other algorithms, at least, not without relying on much more complicated assumptions (my preference requires hash functions, not LWE/MSIS).

**Comment by spirobel** [comment]
>I read the reasoning. IMO, sub-optimal safety thresholds aren't worth the performance benefit.

they are worth it. people will always prefer the fastest solution. It is better to think backwards from the game theoretic end state than sink into perfectionism. 
1. in the case of Monero we want as low a stake as possible in any case to increase the privacy pool
2. safety is just a liveness problem with the added twist that a third party potentially lost money to a double spend. After the network is restarted and rolled back to a state before the attack, the validators can vote to slash the attacker stake or use it to compensate the third party, if this third party is sufficiently likeable. 

> It's that the priority is a safe algorithm comprehensive to asynchronicity, as usable to justify running validators behind networks with extremely-variable latency. 

The thing I want to avoid is that we get stuck on a low market cap NIH half-half hybrid solution that carries baggage and makes finality slower than it needs to be. 

> It's that the priority is a safe algorithm comprehensive to asynchronicity, as usable to justify running validators behind networks with extremely-variable latency.

the danger is that the network is in a half half limbo state where nobody knows if it is still working or not. Alpenglow can tolerate a large percentage of stake being offline. It fits the requirements. It comes from an ecosystem that had to deal with recovering from outages while there was a real market cap and users at stake. Better focus on that instead of theory.

**Comment by binarybaron** [comment]
> > I read the reasoning. IMO, sub-optimal safety thresholds aren't worth the performance benefit.
> 
> they are worth it. people will always prefer the fastest solution. It is better to think backwards from the game theoretic end state than sink into perfectionism.
> 
> 1. in the case of Monero we want as low a stake as possible in any case to increase the privacy pool
> 2. safety is just a liveness problem with the added twist that a third party potentially lost money to a double spend. After the network is restarted and rolled back to a state before the attack, the validators can vote to slash the attacker stake or use it to compensate the third party, if this third party is sufficiently likeable.
> 
> > It's that the priority is a safe algorithm comprehensive to asynchronicity, as usable to justify running validators behind networks with extremely-variable latency.
> 
> The thing I want to avoid is that we get stuck on a low market cap NIH half-half hybrid solution that carries baggage and makes finality slower than it needs to be.
> 
> > It's that the priority is a safe algorithm comprehensive to asynchronicity, as usable to justify running validators behind networks with extremely-variable latency.
> 
> the danger is that the network is in a half half limbo state where nobody knows if it is still working or not. Alpenglow can tolerate a large percentage of stake being offline. It fits the requirements. It comes from an ecosystem that had to deal with recovering from outages while there was a real market cap and users at stake. Better focus on that instead of theory.

I do not think the value proposition of Serai is to be "the fastest cross chain dex".

**Comment by kayabaNerve** [comment]
> in the case of Monero we want as low a stake as possible in any case to increase the privacy pool

This issue isn't about Monero, but the point a low stake increases validators increases decentralization is heard. In that case, I'd cite Asynchronous Algorand which exceeds general performance by introducing an error term, yet largely maintains 3f+1.

> safety is just a liveness problem with the added twist that a third party potentially lost money to a double spend. After the network is restarted and rolled back to a state before the attack, the validators can vote to slash the attacker stake or use it to compensate the third party, if this third party is sufficiently likeable.

Strong disagree, especially here within the context of Serai. Bitcoin sent out because of a finalized chain cannot be recovered. We _need_ strong finality.

> Alpenglow can tolerate a large percentage of stake being offline. 

Alpenglow handles 40% being offline (not 33%) in exchange for not making any progress on networks with extremely-variable latency (unless it simply increases its timeouts with every failure). The marginal benefit in nodes required to be alive isn't worth it if it generally doesn't work without very long timeouts (as the timeouts need to be set to the average worst case, not the best case. In an asynchronous consensus algorithm, the actual state of the network is the rate at which it progresses).

To put this into reality: If validators behind Tor take a minute to execute a consensus protocol once every five minutes, the block time must be set to one minute to be stable. If all other times, it works within a few seconds, a synchronous consensus algorithm will simply sleep for ~57 seconds. An asynchronous network will produce blocks as it can though, meaning a few seconds when a few seconds, and a minute when a minute, at the cost of increased overhead per each block.

**Comment by kayabaNerve** [comment]
Anecdotally, I appreciate Mysticeti for its practical adoption. I also really like AleoBFT's formal verification. My current preference, which is my preference, not my ultimatum, is due to its simplicity and post-quantum properties allowing us to use it for _years_ without risking a mad scramble to throw it entirely away and rebuild it when a QC is on the horizon. Most asynchronous consensus algorithms rely on non-PQ cryptography, and while there are those using PQ assumptions, they'd require we combine random coins from more traditional assumptions or fully trust our choice of LWE/MSIS parameters.

I believe a _baseline_ of future-safe consensus is most important, before moving to further and further optimizations. While we do have flexibility in the algorithm we initially target, I still believe my preferred algorithm is the best baseline.

**Comment by spirobel** [comment]
> Strong disagree, especially here within the context of Serai. Bitcoin sent out because of a finalized chain cannot be recovered. We need strong finality.

The reason you disagree is because in the case of Serai the validators are both counterparties and nodes in the network. You have to split the consensus into two parts: 1. execution of transactions and verification that no double spend happened. 2. the performance of the counterparty role that settles transactions on other networks.

During the first part of the consensus, the fault we need to defend against is denial of service. The default failure case of the Byzantine Generals Problem is not theft. It is denial of service.

Quote from the Byzantine Generals paper: "With unforgeable written messages, the problem is solvable for any number of
generals and possible traitors."
https://lamport.azurewebsites.net/pubs/byz.pdf

We state that a participant in the first part of the consensus has to put up a Generals Bond of 1 Serai (in the case of Monero 1 XMR). If this participant to the network decides to double spend, every counterparty to the network, from the street vendor to the billion dollar exchange will observe it. 
As a result, the network halts as the consensus can't move to the second phase where people act as counter parties to the network.
The network is restarted after the Generals Bond of the perpetrator is wiped out. 

Why did nobody discover this earlier? People have a hard time separating themselves from being a counterparty to the network while participating in the maintenance of its safety. But now that these two roles are separated we can clearly see that we are only defending against liveness failures not theft. 

It is fair to find a new name for this. Note the use of the term Generals Bond as opposed to stake.

**Comment by kayabaNerve** [comment]
It only halts if the equivocation is detected, which requires a synchrony assumption. Allowing smaller and smaller subsets to produce finalizations increases the risk two may simultaneously co-exist until effect.

**Comment by spirobel** [comment]
It is interesting that the response to weightless stake is a comment regarding network topology. 

Lets accept the assumption that there will be a time when we have to choose between liveness and safety according to CAP / FLP impossibility. 
This situation happened to Bitcoin multiple times. One time in 2013 and once in 2010:
2010:
>On August 15 2010, it was discovered that block 74638 contained a transaction that created 184,467,440,737.09551616 bitcoins for three different addresses.[1][2][3] Two addresses received 92.2 billion bitcoins each, and whoever solved the block got an extra 0.01 BTC that did not exist prior to the transaction. This was possible because the code used for checking transactions before including them in a block didn't account for the case of outputs so large that they overflowed when summed.[4]

>A new version of the client was published within five hours of the discovery that contained a soft forking change to the consensus rules that rejected output value overflow transactions (as well as any transaction that paid more than 21 million bitcoins in an output for any reason).[5] The block chain was forked. Although many unpatched nodes continued to build on the "bad" block chain, the "good" block chain overtook it at a block height of 74691[6] at which point all nodes accepted the "good" blockchain as the authoritative source of Bitcoin transaction history. src: https://en.bitcoin.it/wiki/Value_overflow_incident

2013:
>What had happened was the following. The latest version 0.8 release of bitcoind, by far the most popular implementation of Bitcoin used by miners, switched the database that it used to store blocks and transactions from BerkeleyDB to the more efficient LevelDB as part of an effort to reduce blockchain synchronization time. **However, what the developers did not realize at the time was that by doing so they also accidentally introduced a change to the rules of the Bitcoin protocol.** src: https://bitcoinmagazine.com/technical/bitcoin-network-shaken-by-blockchain-fork-1363144448

As a result of this assumption we are not operating under Byzantine conditions when we are not faced with this choice.
Both Proof of Work and Proof of Stake impose a cost on an adversary that wants to force the network into this state. 

There are two negative side effects as a result of this state:
1. potential loss for a counterparty not aware of Byzantine conditions
2. temporary existential crisis of the market during these moments (as it operates under the wrong assumption that the network has permanently solved for surviving the Byzantine state automatically)

There is also potential gain (with or without getting caught) for the attacker: 
1. tricking the unaware counterparty into handing over something valuable
2. shorting the market on potentially KYC free exchanges

PoW and PoS impose this cost on the adversary as a side product of their protocols that pretend to operate under Byzantine conditions. To evaluate if this cost is large enough and the system is stable, as well as to assess if any unnecessary caveats are introduced, lets consider a protocol according to this concept we refer to as Proof of Generals for now (if anyone has a suggestion for a better name, go ahead).
Description of the protocol:
A leader is picked with round robin in the order that the Generals Bond of the nodes was locked. The nodes check if the proposed block is valid and publish the result. We define a finality value as such that we are willing to shut down the network if an attacker pays this price. The nodes will calculate the number of votes times the Generals Bond and verify the block. If the value is equal to or higher than the defined finality value and the block is invalid, the node will terminate.

Once the network is restarted, the leader that proposed the invalid block and all the nodes that voted to verify it as a valid block will have their Generals Bond wiped out.

What is the result of this? 

The network is permissionless. The Generals Bond is to be picked reasonably low, so that anyone who can afford to run a server will be able to afford running a node that actively participates in the networks security. 

There is no ability for an attacker to profit from this, as 
1. Every counterparty that runs the node software will be shut down and be aware of the failure. So it will be impossible for an adversary to trick them into handing over valuables while unaware. 
2. The market is aware that a liveness failure is not a safety failure. The finality value has to be picked high enough, to outweigh the potential to profit from a liveness outage.

How do PoW, PoS or PoW-PoS hybrid solutions compare according to this framework?

According to this [article](https://powerupprivacy.com/2025/03/05/monero-fifty-one-percent-attack-analysis.html) the hardware cost is around 13 million usd in terms of CPUs to perform this attack on Monero. To account for other necessary hardware and electricity we set the finality value to 20 million to construct an equivalent under the Proof of Generals framework.
Under PoW the community has to own at least a similar amount of CPUs and consistently outbid the attacker(if the attacker decides to add more CPUs) to restart the network and keep it running. The PoG system will be free to restart while the attacker loses 20 million.

In PoS systems the necessary stake is often picked in such a way that creates a quasi permissioned network as it prices out most people from realistically running a node. This ties back to the beginning: the stake is not just used to prevent this attack it is also used to create a certain network topology. Rewards for block production create incentives to DDOS leaders and make the security of the system harder to reason about in general.

In the case of hybrid PoW-PoS protocols it really depends on the details and which of the two takes precedence.

**Comment by kayabaNerve** [comment]
While these are not asynchronous consensus algorithms, I want to briefly discuss the current choice of Tendermint for the tributaries. It was chosen as it's a dead-simple consensus protocol of reasonable efficiency. However, in the years since,

- [HotStuff](https://arxiv.org/abs/1803.05069v6), which was not only more efficient, but worked to provide more of a framework for evaluating consensus protocols. This, IMO, is one of the reasons its cited so much more in recent years. On paper, it does require threshold signatures, and possibly could be considered of greater implementation complexity due to the pipelining which occurs when transforming from Basic HotStuff to Chained HotStuff. It also requires the leader not only be online to propose a block but to aggregate threshold signature shares during the round.
- [HotStuff-2](https://eprint.iacr.org/2023/397) further reduced the latency of HotStuff by having proposers broadcast not just the block, but also the certificate they cite. This actually seems approximate to what `tendermint-machine` currently does, broadcasting `ValidRound` with the proposal (though we do it in order to achieve a constant memory bound, not to reduce latency/implement a distinct protocol). It should be noted HotStuff-2 further emphasizes leaders _during the round as a whole_ (as enabling lower communication), where the votes of a block are sent to the _next round's leader_.
- [(Dispersed)Simplex](https://eprint.iacr.org/2023/1916) (which I cite for how it also presents Simplex) may actually be simpler than Tendermint, and simply has the unfortunate result of appearing after our work began. While we could work to upgrade to it, it just isn't worth the time/effort. This is especially since, as presented, it requires a certificate for every round (not just a commit for each block), which is fine but awkward. While there is a [variant which solves this](https://decentralizedthoughts.github.io/2026-04-26-complex/), then we have the complexity of implementing a recent variant of unknown review and further research. Additionally, while discussing variants, we'd likely want to benefit from [C-Simplex](https://decentralizedthoughts.github.io/2025-09-24-Kuplex/) into 'C-Complex' if possible.

**Comment by kayabaNerve** [comment]
It doesn't appear I ever explicitly said this on this issue, but https://eprint.iacr.org/2024/677 is my explicit preference for asynchronous consensus at this time.
